### PR TITLE
CTF: CRHook fixes & Improvements

### PR DESCRIPTION
**Pull request by inf1niti**
This PR cleans up some issues with the existing `crhook` (fast hook) implementation and is based on community discussion & player feedback provided in the QWCTF discord -> [see conversation here](https://discord.com/channels/778331251332743209/966951245627666482/1044917373859266610)

* adds 3 votable hook options: smooth (default), fast, classic
* smooth: adds "cancel time" of 0.5s to initial throw (no spam)
* remove grapple chain sound, improve sound implementation (specify channels)
* fast/smooth: reduce grapple velocity from 1200 to 1050
* smooth: add linear accel/decel to hook velocity changes (no immediate speed change)
* smooth: add fire delay to repeated grappled (no spam)
* remove annoying looped grapple sounds for simpler axe swish(throw) and axe hit (anchor) sounds
* improve balance of haste-hook implementation, make grapple speed a factor of haste power
* move rune sounds to a more appropriate channel (jump no longer cancels sounds)
* decouple hook delay from weapon switch/fire (delay subsequent hooks, not RL/LG/ETC)
* fix chain clear bug, improve hook animation smoothness

### CODE REVIEW

File: src/grapple.c [hunk 1]
```diff
@@ -11,6 +11,25 @@
...
+	return sqrt(
+		velocity[0] * velocity[0] +
+		velocity[1] * velocity[1] +
+		velocity[2] * velocity[2]
+	);
+}
+
+float IncrementSpeed(float speed, float incr)
+{
+	return speed < (800 - incr) ? speed += incr : 800;
```
Comment (tcsabina) [inline src/grapple.c:25]: please make this human readable :D.

```
float retVal = 800;

if (speed < (800 - incr))
{
  retVal = speed + incr;
}

return retVal;
```

or
```
float retVal = speed + incr;
if (retVal > 800)
{
  retVal = 800;
}
```

File: src/grapple.c [hunk 2]
```diff
@@ -11,6 +11,25 @@
...
+}
+
+float DecrementSpeed(float speed, float decr)
+{
+	return speed > (800 + decr) ? speed -= decr : 800;
```
Comment (tcsabina) [inline src/grapple.c:30]: same here. these one liners are super-geek, but it is hard to process them quickly...

File: src/grapple.c [hunk 3]
```diff
@@ -166,16 +185,15 @@ void UpdateChain()
...
 
-	// allow hook to reset mid-throw if clanring hook is enabled
 	if (cvar("k_ctf_cr_hook"))
 	{
-		if (!owner->s.v.button0 && (owner->s.v.weapon == IT_HOOK))
-		{
-			GrappleReset(owner->hook);
+		owner->hook_cancel_time += 1;
+		// delay cancelling the hook until ~500ms (13 * 38) after throw (prevent spam attacks)
+		if (owner->hook_cancel_time > 38) {
```
Comment (tcsabina) [inline src/grapple.c:195]: code style: opening bracket to the new line, please

File: src/grapple.c [hunk 4]
```diff
@@ -198,6 +216,17 @@ void UpdateChain()
 	self->s.v.nextthink = next_frame();
 }
 
+void CancelCheck(gedict_t *owner) {
```
Comment (tcsabina) [inline src/grapple.c:219]: code style: opening bracket to do new line, please

**Review by tcsabina — CHANGES_REQUESTED** [review]
~Could you please add a short 'changelog' as well? So we can add it to the release notes when this is published.~
upps, haven't seen your initial comment ;)

**Comment by inf1niti** [comment]
thanks @tcsabina, adjusted the code as requested.   Would like to get some review from @dsvensson and ideally some testing from the QWCTF group prior to merging this as well.

File: src/grapple.c [hunk 5]
```diff
@@ -11,6 +11,35 @@
...
+		velocity[0] * velocity[0] +
+		velocity[1] * velocity[1] +
+		velocity[2] * velocity[2]
+	);
+}
+
+float IncrementSpeed(float speed, float increment)
+{
+	float newSpeed = 800; 
+	if (speed < (800 - increment))
```
Comment (tcsabina) [inline src/grapple.c:26]: code style:
please tag the function with some newlines:
- after the variable definitions/declarations
- after a block (closing bracket of an if statement)

File: src/grapple.c [hunk 6]
```diff
@@ -11,6 +11,35 @@
...
+	{
+		newSpeed = speed + increment;
+	}
+	return newSpeed;
+}
+
+float DecrementSpeed(float speed, float decrement)
+{
+	float newSpeed = 800; 
+	if (speed > (800 + decrement))
```
Comment (tcsabina) [inline src/grapple.c:36]: same here: please add some newlines as suggested above

File: src/grapple.c [hunk 7]
```diff
@@ -327,37 +362,47 @@ void GrappleService()
 		VectorSubtract(self->hook->s.v.origin, self->s.v.origin, hookVector);
 
 	// No longer going to factor maxspeed into grapple velocity
-	// Using standard grapple velocity of 800 set from original 3wave
 	// purectf velocity = 2.35 * 320 or 360 * 1 = 750 or 846
 
 	VectorCopy(hookVector, hookVelocity);
 	VectorNormalize(hookVelocity);
 
+	if(cvar("k_ctf_cr_hook"))
```
Comment (tcsabina) [inline src/grapple.c:370]: code style: extra space between if an opening parenthesis

**Review by tcsabina — CHANGES_REQUESTED** [review]

**Comment by tcsabina** [comment]
We can probably can summon dsvenson (aka nano from discord), but how do you see the testing?
As the changes only affect the CTF/grapling hook code, I am not afraid to merge it into master, to create a testable snapshot. What do you think? I think that would be easier for a CTF server admin: just get the latest snapshot and update a few server port.

**Comment by dsvensson** [comment]
You could also change the default of k_ctf_rune_power_hst from 2 to 1.25. I think that was an incorrect port from PureCTF Pro (which has 1.25) when it got added. I'll read up on the changes tomorrow.

**Comment by tcsabina** [comment]
ok lads. let me know when to merge!

File: src/grapple.c [hunk 8]
```diff
@@ -327,37 +366,47 @@ void GrappleService()
...
+	else
+	{
+		self->grapple_vel = 800;
+	}
+
+
 	if (self->ctf_flag & CTF_RUNE_HST)
 	{
-		VectorScale(hookVelocity, 800 + (100 * (cvar("k_ctf_rune_power_hst") / 2 + 1)),
+		VectorScale(hookVelocity, self->grapple_vel * 1.125,
```
Comment (dsvensson) [inline src/grapple.c:393]: This takes a step away from traditional haste settings. PureCTF lineage of CTF defaulted to hst power 1.25, thus player moved at maxspeed * 1.25, as well as hook movement * 1.25, and shooting speed * 1.25.

**Comment by dsvensson** [comment]
It's still work in progress. Been doing a mini tournament in the CTF corner that has delayed this a bit. Hope to get some additional testing this week, but game feel is very close to finished imho. Lots of people like the changes. Just minor comments here and there. Hasn't stirred up any huge controversies other than a debatable delay.

**Comment by inf1niti** [comment]
@tcsabina @dsvensson this has been updated.  I've added some new voting logic, so feel free to 🔍 & make suggestions as necessary. 👍 

there is one KTX config change that needs to be propagated when this merges:
**k_ctf_cr_hook:** `0 || 1 (default)` 
becomes 
**k_ctf_hookstyle:** `1 (default) || 2 || 3`

Let me know where/when is best to do that (nquake server?)

**Review by tcsabina — APPROVED** [review]

**Comment by tcsabina** [comment]
@inf1niti 

Could you explain the 3 versions of k_ctf_hookstyle? I would add it to the Release Notes.

And I think you are right about nquakesv: we should tell those guys to update the stuff whenever KTX 1.43 is released.

**Comment by inf1niti** [comment]
> Could you explain the 3 versions of k_ctf_hookstyle? I would add it to the Release Notes.

the `k_ctf_hookstyle` cvar lets players decide between three styles of grappling hook: 

## `k_ctf_hookstyle 1`: smooth hook (default) 
The smooth hook implements some new features to help balance CTF gameplay  
  * throw speed has been reduced from cr_hook (1200 -> 1050), while still being faster than the classic hook (800)  
  * player acceleration from the hook pull runs through an easing function, giving the hook an overall "smoother" feel
  * implements a fire delay, limiting "hook spam" approach, encouraging more common movement mechanics (bhop).
  * the hook throw is cancelable, after a short delay.
  
## `k_ctf_hookstyle 2`: fast hook
the fast hook is closest to the existing cr_hook implementation, but the speed has been reduced based on community feedback
  * uses the same speed settings as smooth hook, but does not use hook acceleration, or fire delay.
  * the hook throw is immediately cancelable.
  
## `k_ctf_hookstyle 3`: **classic hook** 
the classic hook uses the original 3wave settings
  * hook throw speed is slower (800)
  * no acceleration / easing function to hook pull
  * a hook throw cannot be canceled until the grapple hits a surface

**Comment by tcsabina** [comment]
> > Could you explain the 3 versions of k_ctf_hookstyle? I would add it to the Release Notes.
> 
> the `k_ctf_hookstyle` cvar lets players decide between three styles of grappling hook.
> ## `k_ctf_hookstyle 1`: smooth hook (default)
> 
> _votable with `hook_smooth`, 51% player vote threshold required_
> 
> The smooth hook implements some new features to help balance CTF gameplay
> 
>     * throw speed has been reduced from cr_hook (1200 -> 1050), while still being faster than the classic hook (800)
> 
>     * player acceleration from the hook pull runs through an easing function, giving the hook an overall "smoother" feel
> 
>     * implements a fire delay, limiting "hook spam" approach, encouraging more common movement mechanics (bhop)
> 
>     * the hook throw is cancelable, after a short delay
> 
> 
> ## `k_ctf_hookstyle 2`: fast hook
> 
> _votable with `hook_fast`, 51% player vote threshold required_
> 
> the fast hook is closest to the existing cr_hook implementation, but the speed has been reduced based on community feedback
> 
>     * uses the same speed settings as smooth hook
> 
>     * no hook pull acceleration/easing, or fire delay
> 
>     * the hook throw is immediately cancelable
> 
> 
> ## `k_ctf_hookstyle 3`: **classic hook**
> 
> _votable with `hook_classic`, 51% player vote threshold required_
> 
> the classic hook uses the original 3wave settings
> 
>     * hook throw speed is slower (800)
> 
>     * no hook pull acceleration/easing, or fire delay
> 
>     * a hook throw cannot be canceled until the grapple hits a surface

Perfect!

**Comment by mushis** [comment]
I would like to see more information (description) when chooseing hook mode.
the description of the different hook modes should have more info. example:
`hook_smooth -> "reduced throw speed, with fire delay and acceleration"`

**Comment by tcsabina** [comment]
> I would like to see more information (description) when chooseing hook mode. the description of the different hook modes should have more info. example: `hook_smooth -> "normal throw speed, with fire delay and acceleration"` `hook_classic -> "slower throw speed, no accel, not cancellable"` `hook_fast -> "normal throw speed, no accel, no delay"`

So read to comment from inf1niti above ;)

**Comment by mushis** [comment]
> So read to comment from inf1niti above ;)

I meant in the game =) 
as a description of the commands.

**Pull request merged**